tests/session/test_env_select.py

```diff
@@ -15,6 +15,57 @@
 
 CURRENT_PY_ENV = f"py{sys.version_info[0]}{sys.version_info[1]}"  # e.g. py310
 
+####################################################################
+#   Tests directly on env_select functionality.
+
+
+@pytest.mark.parametrize(
+    ("user_input", "env_names", "is_all", "is_default"),
+    [
+        (None, (), False, True),
+        ("", (), False, True),
+        ("a1", ("a1",), False, False),
+        ("a1,b2,c3", ("a1", "b2", "c3"), False, False),
+        #   If the user gives "ALL" as any envname, this becomes an "is_all" and other envnames are ignored.
+        ("ALL", (), True, False),
+        ("a1,ALL,b2", (), True, False),
+        #   Zero-length envnames are ignored as being not present. This is not intentional.
+        (",,a1,,,b2,,", ("a1", "b2"), False, False),
+        (",,", (), False, True),
+        #   Environment names with "invalid" characters are accepted here; the client is expected to deal with this.
+        ("\x01.-@\x02,xxx", ("\x01.-@\x02", "xxx"), False, False),
+    ],
+)
+def test_clienv(user_input: str, env_names: tuple[str], is_all: bool, is_default: bool) -> None:
+    c = CliEnv(user_input)
+    assert (c.is_all, c.is_default_list, tuple(c)) == (is_all, is_default, tuple(env_names))
+
+
+@pytest.mark.parametrize(
+    ("user_input", "expected"),
+    [
+        ("", False),
+        ("all", False),
+        ("All", False),
+        ("ALL", True),
+        ("a,ALL,b", True),
+    ],
+)
+def test_clienv_is_all(user_input: str, expected: bool) -> None:
+    assert CliEnv(user_input).is_all is expected
+
+
+def test_env_select_lazily_looks_at_envs() -> None:
+    state = State(get_options(), [])
+    env_selector = EnvSelector(state)
+    # late-assigning env should be reflected in env_selector
+    state.conf.options.env = CliEnv("py")
+    assert set(env_selector.iter()) == {"py"}
+
+
+####################################################################
+#   Tests of env_select functionality as used through other code.
```

Comment:
I'm -1 on adding these markers 🤔

Reply:
Will be resolved with resolution of lines 18-19.